### DISCUSSION TITLE
Support for different types of ListItem

### DISCUSSION THREAD

**Discussion by chapuzzo** [ueberdosis/tiptap#671] [General Questions]
I've been looking through all of the documentation but couldn't find anything related to `li` types.
Currently the only supported types are the default ones, I see these are the available ones: 

for ol:
- 1: It is the default value. It is used to specify the numerical ordered list.
- a: It arranged the list items in lower case letters.
- A: It arrange the list items in the form of upper case.
- i: It arrange the list items in the roman numbers in the form of lower case letters.
- I: It arranged the list in roman numerals in the form of uppercase letters.

for ul:
- disc: It is the default value. It creates a filled circle.
- circle: It creates an unfilled circle.
- square: It creates an filled square.

Is there any option to change them or a plan for this to be enabled?

**Comment by philippkuehn** [comment]
I don't understand. Isn't that just CSS?

**Reply by ashu12chi** [reply]
The CSS will not be helpful for everyone, I also want this feature and logged #3545. I explained in the issue, why this is not useful for me, Quoting from that issue:

> The issue with styling via CSS is type will not be exported as part of exported HTML, which is required for my use case. Additionally, when I copy content from an external source like MS Outlook, I want to parse the type attribute to have the exact same styling.

**Comment by chapuzzo** [comment]
Yes and no. They can be set by CSS but they are also a deprecated attribute https://developer.mozilla.org/en-US/docs/Web/HTML/Element/li 😞 
How could they have custom CSS just like the heading (1,2,3) examples?

**Comment by philson-philip** [comment]
@philippkuehn @chapuzzo Inside [TinyMCE](https://www.tiny.cloud/) editor, they support multiple `list-style-type` to both ordered list and unordered list. While working on different projects, I was blocked by the same requirement and was forced to use TinyMCE editor instead of tiptap editor. Can we have a solution for this?

**Reply by philson-philip** [reply]
<img width="398" alt="Screenshot 2023-01-24 at 9 43 38 AM" src="https://user-images.githubusercontent.com/22117214/214210935-302b01bd-6226-49d3-9ce8-0a82d183fbb2.png">

<img width="274" alt="Screenshot 2023-01-24 at 9 43 59 AM" src="https://user-images.githubusercontent.com/22117214/214210961-dda238b2-41a5-440d-a4a2-7aa9bd9419f6.png">

**Reply by katerlouis** [reply]
Have you found a solution for that? I also want to provide my editors different types of unordered lists and don't know how to best achieve that.

**Reply by philson-philip** [reply]
@katerlouis I had to use TinyMCE editor instead of tiptap to resolve this issue. Waiting for the team to create an issue for this.

**Comment by sangeethkc** [comment]
@chapuzzo @katerlouis  Did you guys find any solution to this issue? If yes, please update.

**Comment by whykhamist** [comment]
For anyone still looking, here's what I did 

First I extended the BulletList to add the "class" attribute then a custom function to toggle class names ( toggleClass(...) ). I used tailwindcss to change the list-style-type. 

Caveats:
- The custom attribute and command becomes available to OrderedList as well even without extending it. 
- using the custom function on the OrderedList will convert it to a BulletList
- changing the style of a nested list will also update its parent list.
- isActive is not exclusive to the current node, It is affected by the parent as well.
![image](https://github.com/ueberdosis/tiptap/assets/29524439/75409720-c30b-4cdc-a482-c74f0153102e)

```js
const BulletListCustom = BulletList.extend({
    addAttributes() {
      return {
        class: {
          default: "list-disc",
        },
      };
    },
    addCommands() {
      return {
        ...this.parent?.(),
        toggleClass:
          (className) =>
          ({ commands, chain }) => {
            if (!this.editor.isActive(this.name)) {
              return chain().toggleBulletList().updateAttributes(this.name, {
                class: className,
              });
            }

            if (!this.editor.isActive(this.name, { class: className })) {
              return commands.updateAttributes(this.name, { class: className });
            }
          },
      };
    },
  });
```

And to use it:

```html
<YourButton
    class="aspect-square w-10"
    icon="square"
    :active="editor?.isActive('bulletList', { class: 'list-[square]' })"
    @click="editor?.chain().focus().toggleClass('list-[square]').run()"
/>
```

**Reply by richardstowey** [reply]
I can't seem to get your solution working. The bullet lists still don't have any styling, and toggling the styles off doesn't work either.

I don't have an answer to suggest, but this is still very much an issue for me, and I'm not sure why this doesn't work. It should be fundamental on the list of requirements.

**Reply by whykhamist** [reply]
Here's a working [demo](https://codesandbox.io/p/github/whykhamist/tiptap-custom-bullet/main?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clsdrvdw10006356fdp3e2tak%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clsdrvdw00002356fzpz3gco9%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clsdrvdw00004356fdjgl6vd7%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clsdrvdw00005356fjvt1rpcv%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clsdrvdw00002356fzpz3gco9%2522%253A%257B%2522id%2522%253A%2522clsdrvdw00002356fzpz3gco9%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522clsdrvdw00005356fjvt1rpcv%2522%253A%257B%2522id%2522%253A%2522clsdrvdw00005356fjvt1rpcv%2522%252C%2522activeTabId%2522%253A%2522clsdwjdsj00kr356fgh7sy0yl%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A5173%252C%2522id%2522%253A%2522clsdwjdsj00kr356fgh7sy0yl%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%255D%257D%252C%2522clsdrvdw00004356fdjgl6vd7%2522%253A%257B%2522id%2522%253A%2522clsdrvdw00004356fdjgl6vd7%2522%252C%2522activeTabId%2522%253A%2522clsdrvfqv005p356fhoji9vkw%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clsdrvdw00003356f0tkxunuw%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clsdrvfq9000sdkhkbisp12r6%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522id%2522%253A%2522clsdrvfqv005p356fhoji9vkw%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522CSB_RUN_OUTSIDE_CONTAINER%253D1%2520devcontainer%2520templates%2520apply%2520--template-id%2520%255C%2522ghcr.io%252Fdevcontainers%252Ftemplates%252Fjavascript-node%255C%2522%2520--template-args%2520%27%257B%257D%27%2520--features%2520%27%255B%255D%27%2522%252C%2522id%2522%253A%2522clsdwimf60071356fsc7j7et3%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D) using Vue3.

Try to inspect the node if the class name is there, if it is, you may need to add `!important` to the style. I believe the `toggleClass(...)` function can yet be enhanced but for now, this works for me.

**Reply by richardstowey** [reply]
Thanks @whykhamist , I managed to get it working with your example!
